forum/forum/logging.py

```diff
@@ -0,0 +1,37 @@
+import logging
+from typing import Literal
+
+
+class ColoredFormatter(logging.Formatter):
+    grey = "\x1b[38;20m"
+    green = "\u001b[32m"
+    yellow = "\x1b[33;20m"
+    red = "\x1b[31;20m"
+    magenta = "\u001b[35m"
+    reset = "\x1b[0m"
+
+    def __init__(
+        self,
+        fmt: str | None = None,
+        datefmt: str | None = None,
+        style: Literal['%'] | Literal['{'] | Literal['$'] = "%",
+        validate: bool = True,
+        *,
+        defaults=None
+    ) -> None:
+        super().__init__(fmt, datefmt, style, validate, defaults=defaults)
+        self.FORMATS = {
+            logging.DEBUG: self.grey + self._fmt + self.reset,
+            logging.INFO: self.green + self._fmt + self.reset,
+            logging.WARNING: self.yellow + self._fmt + self.reset,
+            logging.ERROR: self.red + self._fmt + self.reset,
+            logging.CRITICAL: self.magenta + self._fmt + self.reset
+        }
```

Comment:
Precompute the log formats using the provided format string `fmt`:
```
        super().__init__(fmt, datefmt, style, validate, defaults=defaults)
        self.FORMATS = {
            logging.DEBUG: self.grey + (fmt or self._fmt) + self.reset,
            logging.INFO: self.green + (fmt or self._fmt) + self.reset,
            logging.WARNING: self.yellow + (fmt or self._fmt) + self.reset,
            logging.ERROR: self.red + (fmt or self._fmt) + self.reset,
            logging.CRITICAL: self.magenta + (fmt or self._fmt) + self.reset
        }
```

Reply:
Is it really necessary?

This is `__init__` method of `Formatter`
```Python
    def __init__(self, fmt=None, datefmt=None, style='%', validate=True, *,
                 defaults=None):

        if style not in _STYLES:
            raise ValueError('Style must be one of: %s' % ','.join(
                             _STYLES.keys()))
        self._style = _STYLES[style][0](fmt, defaults=defaults)
        if validate:
            self._style.validate()

        self._fmt = self._style._fmt
        self.datefmt = datefmt
```

`self._fmt = self._style._fmt`

So it seems that we can omit such instruction.